### PR TITLE
Autoscaler integrates with operator by the node provider

### PR DESCRIPTION
**Pull request by Qstar**
## Why are these changes needed?

1.fix the problem falling behind in terms of scaling features; autoscaling not up to par with other Clouds

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/_private/kubernetes_operator/ray_cluster_util.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+"""The Ray Operator Node Provider utils."""
```
Comment (caitengwei) [inline python/ray/autoscaler/_private/kubernetes_operator/ray_cluster_util.py:1]: It's a constant file, not utility file.

File: deploy/ray-operator/api/v1alpha1/raycluster_types.go [hunk 2]
```diff
@@ -86,21 +86,30 @@ type Extension struct {
...
+	IdList []string `json:"idList"`
+
 	// Logical groupName for worker in same group, it's used for heterogeneous feature to distinguish different groups.
 	// GroupName is the unique identifier for the pods with the same configuration in one Ray Cluster
 	GroupName string `json:"groupName,omitempty"`
 
 	// Command to start ray
 	Command string `json:"command,omitempty"`
 
+	// Args to start ray
```
Comment (caitengwei) [inline deploy/ray-operator/api/v1alpha1/raycluster_types.go:99]: Please explain why `args` is needed along with `command`. Provide an example in comment will be nice.

File: deploy/ray-operator/api/v1alpha1/raycluster_types.go [hunk 3]
```diff
@@ -86,21 +86,30 @@ type Extension struct {
...
 	// Labels are intended to be used to specify identifying attributes of objects that are meaningful and relevant to users,
 	// but do not directly imply semantics to the core system. Labels can be used to organize and to select subsets of objects.
 	Labels map[string]string `json:"labels,omitempty"`
 
-	// The service acccount name.
+	// The service account name.
 	ServiceAccountName string `json:"serviceAccountName,omitempty"`
 
+	// The container ports
+	Ports []v1.ContainerPort `json:"ports,omitempty" patchStrategy:"merge" patchMergeKey:"containerPort" protobuf:"bytes,6,rep,name=ports"`
```
Comment (caitengwei) [inline deploy/ray-operator/api/v1alpha1/raycluster_types.go:111]: In out future plan, we can also support this configuration in template, right?

File: deploy/ray-operator/config/crd/bases/ray.io_rayclusters.yaml [hunk 4]
```diff
@@ -671,6 +671,9 @@ spec:
                       be preserved when modifying objects. Optional. Usage refers
                       to https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/
                     type: object
+                  args:
+                    description: Args to start ray
```
Comment (caitengwei) [inline deploy/ray-operator/config/crd/bases/ray.io_rayclusters.yaml:675]: Same question like above, how to distinguish `args` from `command`?

File: deploy/ray-operator/config/crd/bases/ray.io_rayclusters.yaml [hunk 5]
```diff
@@ -791,6 +794,12 @@ spec:
                     description: 'Head service suffix. So head can be accessed by
                       domain name: {namespace}.svc , follows Kubernetes standard.'
                     type: string
+                  idList:
+                    description: create operation should have no idList (generate
```
Comment (caitengwei) [inline deploy/ray-operator/config/crd/bases/ray.io_rayclusters.yaml:798]: Explain more about what idList is designed for.

File: deploy/ray-operator/config/default/kustomization.yaml [hunk 6]
```diff
@@ -1,5 +1,5 @@
 # Adds namespace to all resources.
-namespace: ray-operator-system
+namespace: ray
```
Comment (caitengwei) [inline deploy/ray-operator/config/default/kustomization.yaml:2]: Why change this?

File: deploy/ray-operator/controllers/common/pod.go [hunk 7]
```diff
@@ -85,15 +85,6 @@ func buildContainer(conf *PodConfig) corev1.Container {
...
-			{
-				ContainerPort: int32(defaultRedisPort),
-				Name:          "redis",
-			},
-			{
-				ContainerPort: int32(defaultHTTPServerPort),
-				Name:          "http-server",
-			},
-		},
+		Ports:           conf.Extension.Ports,
```
Comment (caitengwei) [inline deploy/ray-operator/controllers/common/pod.go:88]: Will `Ports` have a default configuration if user didn't provide them?

File: python/ray/autoscaler/_private/kubernetes_operator/node_provider.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+
+        return [pod.metadata.name for pod in pod_list.items]
+
+    def is_running(self, node_id):
+        pod = core_api().read_namespaced_pod(node_id, self.namespace)
+        return pod.status.phase == "Running"
+
+    def is_terminated(self, node_id):
+        pod = core_api().read_namespaced_pod(node_id, self.namespace)
+        return pod.status.phase not in ["Running", "Pending"]
```
Comment (caitengwei) [inline python/ray/autoscaler/_private/kubernetes_operator/node_provider.py:58]: Is `Init` also one of states of non_terminated?

File: python/ray/autoscaler/_private/kubernetes_operator/node_provider.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+
+    def create_node(self, node_config, tags, count):
+        # create a raycluster
+        conf = node_config.copy()
+        pod_spec = conf.get("pod", conf)
+        service_spec = conf.get("service")
+        ingress_spec = conf.get("ingress")
+        node_uuid = str(uuid4())
+        tags[TAG_RAY_CLUSTER_NAME] = self.cluster_name
+        tags["ray-node-uuid"] = node_uuid
```
Comment (caitengwei) [inline python/ray/autoscaler/_private/kubernetes_operator/node_provider.py:89]: Why not use a constant to place "ray-node-uuid"

File: python/ray/autoscaler/_private/kubernetes_operator/node_provider.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+        body['kind'] = CRD_RAY_KIND
+        body['metadata'] = {}
+        body['metadata']['name'] = self.cluster_name
+        spec['clusterName'] = self.cluster_name
+        spec['images'] = {}
+        spec['images']['defaultImage'] = pod_spec['spec']['containers'][0]['image']
+        spec['imagePullPolicy'] = 'Always'
+        return spec
+
+    def generate_extension(self, count, pod_spec, tags, new_node_names):
```
Comment (caitengwei) [inline python/ray/autoscaler/_private/kubernetes_operator/node_provider.py:172]: `generate_extension` is kind of fuzzy, can we use a more specific method name?

**Review by caitengwei — CHANGES_REQUESTED** [review]

File: deploy/ray-operator/api/v1alpha1/raycluster_types.go [hunk 11]
```diff
@@ -86,21 +86,30 @@ type Extension struct {
 	// Pod docker image
 	Image string `json:"image,omitempty"`
 
+	//create operation should have no idList (generate by controller)
+	IdList []string `json:"idList"`
+
```
Comment (chenk008) [inline deploy/ray-operator/api/v1alpha1/raycluster_types.go:91]: Bump the CRD version if we upgrade the struct

File: deploy/ray-operator/api/v1alpha1/raycluster_types.go (hunk 3 above)
Comment (Qstar) [reply deploy/ray-operator/api/v1alpha1/raycluster_types.go:111]: Yes, we will allow users to set in the template.

File: deploy/ray-operator/config/default/kustomization.yaml (hunk 6 above)
Comment (Qstar) [reply deploy/ray-operator/config/default/kustomization.yaml:2]: The k8s node provider creates all the stuff(like serviceAccount) in namespace ray, for users are easy to use, just apply the k8s operator in the same namespace to reduce usage cost.

File: python/ray/autoscaler/_private/kubernetes_operator/node_provider.py (hunk 8 above)
Comment (Qstar) [reply python/ray/autoscaler/_private/kubernetes_operator/node_provider.py:58]: - There are five possible phase values:  Pending, Running, Succeeded, Failed, Unknown.
- Pending: The pod has been accepted by the Kubernetes system, but one or more of the container images has not been created. This includes time before being scheduled as well as time spent downloading images over the network, which could take a while.

File: python/ray/autoscaler/_private/kubernetes_operator/node_provider.py (hunk 9 above)
Comment (Qstar) [reply python/ray/autoscaler/_private/kubernetes_operator/node_provider.py:89]: This rule follows the way in the other node provider implementation.

File: python/ray/autoscaler/_private/kubernetes_operator/ray_cluster_util.py (hunk 1 above)
Comment (Qstar) [reply python/ray/autoscaler/_private/kubernetes_operator/ray_cluster_util.py]: updated

File: deploy/ray-operator/api/v1alpha1/raycluster_types.go (hunk 2 above)
Comment (Qstar) [reply deploy/ray-operator/api/v1alpha1/raycluster_types.go]: Remove the args, it's not used

File: deploy/ray-operator/config/crd/bases/ray.io_rayclusters.yaml (hunk 4 above)
Comment (Qstar) [reply deploy/ray-operator/config/crd/bases/ray.io_rayclusters.yaml]: Removed as not used.

File: deploy/ray-operator/config/crd/bases/ray.io_rayclusters.yaml (hunk 5 above)
Comment (Qstar) [reply deploy/ray-operator/config/crd/bases/ray.io_rayclusters.yaml]: Added comments in the latest commit.

File: deploy/ray-operator/controllers/common/pod.go (hunk 7 above)
Comment (Qstar) [reply deploy/ray-operator/controllers/common/pod.go]: Added default ports when not provided.

File: python/ray/autoscaler/_private/kubernetes_operator/node_provider.py (hunk 10 above)
Comment (Qstar) [reply python/ray/autoscaler/_private/kubernetes_operator/node_provider.py]: Updated the function name.

File: deploy/ray-operator/api/v1alpha1/raycluster_types.go (hunk 11 above)
Comment (Qstar) [reply deploy/ray-operator/api/v1alpha1/raycluster_types.go:94]: This field added as optional for users, do we also update the version?

File: deploy/ray-operator/controllers/common/pod.go [hunk 12]
```diff
@@ -76,6 +76,13 @@ func buildContainer(conf *PodConfig) corev1.Container {
 		corev1.EnvVar{Name: "MY_POD_IP", ValueFrom: &corev1.EnvVarSource{FieldRef: &corev1.ObjectFieldSelector{FieldPath: "status.podIP"}}},
 	)
 
+	var ports []corev1.ContainerPort
+    if conf.Extension.Ports != nil && len(conf.Extension.Ports) >0 {
+        ports = conf.Extension.Ports
+    } else {
+        ports = containerDefaultPorts()
+    }
```
Comment (caitengwei) [inline deploy/ray-operator/controllers/common/pod.go:84]: Need to do some formating

File: deploy/ray-operator/api/v1alpha1/raycluster_types.go (hunk 11 above)
Comment (caitengwei) [reply deploy/ray-operator/api/v1alpha1/raycluster_types.go:94]: I also agree this is necessary

File: deploy/ray-operator/controllers/raycluster_controller.go [hunk 13]
```diff
@@ -179,33 +180,59 @@ func (r *RayClusterReconciler) Reconcile(request reconcile.Request) (reconcile.R
...
+                }
+            } else {
+                var i int32 = 0
+                for i = 0; i < *extension.Replicas; i++ {
+                    podType := fmt.Sprintf("%v", extension.Type)
+                    podName := ""
+                    if extension.GroupName == "" {
+                        podName = instance.Name + common.DashSymbol + extension.GroupName + common.DashSymbol + podType + common.DashSymbol + uuid.New()[0:6]
+                    } else {
+                        podName = instance.Name + common.DashSymbol + podType + common.DashSymbol + uuid.New()[0:6]
```
Comment (caitengwei) [inline deploy/ray-operator/controllers/raycluster_controller.go:208]: Use a utility function to shorten text concat

**Review by caitengwei — CHANGES_REQUESTED** [review]

File: deploy/ray-operator/controllers/common/pod.go (hunk 12 above)
Comment (Qstar) [reply deploy/ray-operator/controllers/common/pod.go]: Formatted in the latest commit.

File: deploy/ray-operator/controllers/raycluster_controller.go (hunk 13 above)
Comment (Qstar) [reply deploy/ray-operator/controllers/raycluster_controller.go]: Extract the method to generate the podName at the latest version.

**Pull request closed without merging**